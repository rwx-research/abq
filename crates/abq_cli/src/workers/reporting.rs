use abq_reporting::{CompletedSummary, Reporter, ReportingError};
use abq_utils::{
    log_assert,
    net_protocol::{client::ReportedResult, queue::AssociatedTestResults},
    results_handler::{NotifyResults, SharedNotifyResults, SharedResultsHandler},
};
use async_trait::async_trait;
use tokio::{sync::mpsc, task::JoinHandle};

use crate::reporting::{reporter_from_kind, ReporterKind, StdoutPreferences, SuiteResult};

use super::summary;

struct ReportingTask {
    rx_results: mpsc::UnboundedReceiver<Vec<AssociatedTestResults>>,
    reporters: Vec<Box<dyn Reporter>>,
    overall_tracker: summary::SuiteTracker,
}

pub(crate) struct FinalizedReporters {
    suite_result: SuiteResult,
    reporters: Vec<Box<dyn Reporter>>,
}

impl ReportingTask {
    fn new(
        rx_results: mpsc::UnboundedReceiver<Vec<AssociatedTestResults>>,
        reporters: Vec<Box<dyn Reporter>>,
    ) -> Self {
        Self {
            reporters,
            rx_results,
            overall_tracker: Default::default(),
        }
    }

    async fn work_forever(mut self) -> FinalizedReporters {
        while let Some(results) = self.rx_results.recv().await {
            results
                .into_iter()
                .for_each(|results| self.handle_associated_results(results));
        }

        self.reporters
            .iter_mut()
            .for_each(|reporter| reporter.after_all_results());

        FinalizedReporters {
            suite_result: self.overall_tracker.suite_result(),
            reporters: self.reporters,
        }
    }

    fn handle_associated_results(&mut self, results: AssociatedTestResults) {
        let AssociatedTestResults {
            work_id: _,
            run_number,
            results,
            before_any_test,
            after_all_tests,
        } = results;

        assert!(
            !results.is_empty(),
            "ABQ protocol must never ship empty list of test results"
        );

        let mut output_before = Some(before_any_test);
        let mut output_after = after_all_tests;

        let mut results = results.into_iter().peekable();

        while let Some(test_result) = results.next() {
            self.overall_tracker
                .account_result(run_number, &test_result);

            let output_before = std::mem::take(&mut output_before);
            let output_after = if results.peek().is_none() {
                // This is the last test result, the output-after is associated
                // with it
                std::mem::take(&mut output_after)
            } else {
                None
            };

            let reported_result = ReportedResult {
                output_before,
                test_result,
                output_after,
            };

            for reporter in self.reporters.iter_mut() {
                // TODO: is there a reasonable way to surface the error?
                let _opt_error = reporter.push_result(run_number, &reported_result);
            }
        }
    }
}

impl FinalizedReporters {
    pub fn finish(self, summary: &CompletedSummary) -> (SuiteResult, Vec<ReportingError>) {
        let Self {
            suite_result,
            reporters,
        } = self;

        let errors: Vec<_> = reporters
            .into_iter()
            .filter_map(|reporter| reporter.finish(summary).err())
            .collect();

        (suite_result, errors)
    }
}

pub(crate) struct ReportingTaskHandle {
    handle: JoinHandle<FinalizedReporters>,
}

impl ReportingTaskHandle {
    pub async fn join(self) -> FinalizedReporters {
        self.handle
            .await
            .expect("reporting task must not be dropped before its handle is")
    }
}

#[derive(Clone)]
pub struct ReportingProxy {
    tx_results: mpsc::UnboundedSender<Vec<AssociatedTestResults>>,
}

#[async_trait]
impl NotifyResults for ReportingProxy {
    async fn send_results(&mut self, results: Vec<AssociatedTestResults>) {
        let res = self.tx_results.send(results);
        log_assert!(
            res.is_ok(),
            "reporting receiver died before all test results consumed"
        );
    }
}

impl SharedNotifyResults for ReportingProxy {
    fn boxed_clone(&self) -> SharedResultsHandler {
        Box::new(self.clone())
    }
}

pub fn build_reporters(
    reporter_kinds: Vec<ReporterKind>,
    stdout_preferences: StdoutPreferences,
    test_suite_name: &str,
) -> Vec<Box<dyn Reporter>> {
    reporter_kinds
        .into_iter()
        .map(|kind| reporter_from_kind(kind, stdout_preferences, test_suite_name))
        .collect()
}

/// Spawns a task to handle reporting of results.
/// Returns a [ReportingProxy] that can be used to communicate with the reporting task.
pub(crate) fn create_reporting_task(
    reporters: Vec<Box<dyn Reporter>>,
) -> (ReportingProxy, ReportingTaskHandle) {
    // TODO make the channel bounded, though in practice this is unlikely to ever materially
    // block since processing results is CPU bound.
    let (tx_results, rx_results) = mpsc::unbounded_channel();

    let reporting_task = ReportingTask::new(rx_results, reporters);
    let task_handle = tokio::spawn(reporting_task.work_forever());

    let proxy = ReportingProxy { tx_results };
    let handle = ReportingTaskHandle {
        handle: task_handle,
    };

    (proxy, handle)
}
