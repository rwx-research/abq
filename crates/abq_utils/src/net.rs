//! Server/client interfaces for use among entities in the ABQ ecosystem.

pub(crate) mod tcp;
#[cfg(not(feature = "tls"))]
pub type ServerListener = tcp::ServerListener;
#[cfg(not(feature = "tls"))]
pub type ServerStream = tcp::Stream;
#[cfg(not(feature = "tls"))]
pub type ConfiguredClient = tcp::ConfiguredClient;
#[cfg(not(feature = "tls"))]
pub type ClientStream = tcp::Stream;

#[cfg(feature = "tls")]
pub(crate) mod tls;
#[cfg(feature = "tls")]
pub type ServerListener = tls::ServerListener;
#[cfg(feature = "tls")]
pub type ServerStream = tls::ServerStream;
#[cfg(feature = "tls")]
pub type ConfiguredClient = tls::ConfiguredClient;
#[cfg(feature = "tls")]
pub type ClientStream = tls::ClientStream;
