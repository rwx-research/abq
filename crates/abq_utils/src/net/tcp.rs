//! Plain old raw TCP connections.
//!
//! A thin shell over the standard libraries' implementations.

use std::{
    io::{self, Read, Write},
    net::{SocketAddr, ToSocketAddrs},
};

pub struct Stream(std::net::TcpStream);

impl Stream {
    pub fn connect(addr: impl ToSocketAddrs) -> io::Result<Self> {
        std::net::TcpStream::connect(addr).map(Self)
    }

    pub fn set_nonblocking(&self, nonblocking: bool) -> io::Result<()> {
        self.0.set_nonblocking(nonblocking)
    }
}

impl Read for Stream {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        self.0.read(buf)
    }
}

impl Write for Stream {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        self.0.write(buf)
    }

    fn flush(&mut self) -> io::Result<()> {
        self.0.flush()
    }
}

pub struct ServerListener {
    pub(crate) listener: std::net::TcpListener,
}

impl ServerListener {
    pub fn bind(addr: impl ToSocketAddrs) -> io::Result<Self> {
        let listener = std::net::TcpListener::bind(addr)?;
        Ok(Self { listener })
    }

    pub fn set_nonblocking(&self, nonblocking: bool) -> io::Result<()> {
        self.listener.set_nonblocking(nonblocking)
    }

    pub fn local_addr(&self) -> io::Result<SocketAddr> {
        self.listener.local_addr()
    }

    pub fn accept(&self) -> io::Result<(Stream, SocketAddr)> {
        let (stream, addr) = self.listener.accept()?;
        Ok((Stream(stream), addr))
    }
}

#[derive(Clone, Copy)]
pub struct ConfiguredClient;

impl ConfiguredClient {
    pub fn new() -> io::Result<Self> {
        Ok(ConfiguredClient)
    }

    pub fn connect(&self, addr: SocketAddr) -> io::Result<Stream> {
        Stream::connect(addr)
    }
}
