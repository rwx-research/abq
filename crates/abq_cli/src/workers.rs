use std::num::{NonZeroU64, NonZeroUsize};
use std::path::PathBuf;
use std::time::Duration;

use abq_reporting::{CompletedSummary, Reporter};
use abq_utils::exit::ExitCode;
use abq_utils::net_protocol::entity::{RunnerMeta, WorkerTag};
use abq_utils::net_protocol::queue::InvokeWork;
use abq_utils::net_protocol::runners::CapturedOutput;
use abq_utils::net_protocol::workers::{RunId, RunnerKind};
use abq_workers::negotiate::{
    NegotiatedWorkers, QueueNegotiatorHandle, WorkersConfig, WorkersNegotiator,
};
use abq_workers::workers::{WorkerContext, WorkersExit, WorkersExitStatus};

mod reporting;
mod retry_manifest_tracker;
mod summary;

use futures::stream::StreamExt;
use signal_hook::consts::TERM_SIGNALS;
use signal_hook_tokio::Signals;

use crate::reporting::{ReporterKind, StdoutPreferences};
use crate::workers::reporting::create_reporting_task;

use self::reporting::{build_reporters, ReportingTaskHandle};

type ClientOptions = abq_utils::net_opt::ClientOptions<abq_utils::auth::User>;

pub fn start_workers_standalone(
    run_id: RunId,
    tag: WorkerTag,
    num_workers: NonZeroUsize,
    runner_kind: RunnerKind,
    working_dir: PathBuf,
    reporter_kinds: Vec<ReporterKind>,
    stdout_preferences: StdoutPreferences,
    batch_size: NonZeroU64,
    test_results_timeout: Duration,
    queue_negotiator: QueueNegotiatorHandle,
    client_opts: ClientOptions,
) -> ! {
    let rt = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    let test_suite_name = "suite"; // TODO: determine this correctly
    let reporters = build_reporters(reporter_kinds, stdout_preferences, test_suite_name);

    rt.block_on(start_standalone_help(
        run_id,
        tag,
        num_workers,
        runner_kind,
        working_dir,
        reporters,
        stdout_preferences,
        batch_size,
        test_results_timeout,
        queue_negotiator,
        client_opts,
    ))
}

async fn start_standalone_help(
    run_id: RunId,
    tag: WorkerTag,
    num_workers: NonZeroUsize,
    runner_kind: RunnerKind,
    working_dir: PathBuf,
    reporters: Vec<Box<dyn Reporter>>,
    stdout_preferences: StdoutPreferences,
    batch_size: NonZeroU64,
    test_results_timeout: Duration,
    queue_negotiator: QueueNegotiatorHandle,
    client_opts: ClientOptions,
) -> ! {
    let mut term_signals = Signals::new(TERM_SIGNALS).unwrap();

    let context = WorkerContext::AlwaysWorkIn { working_dir };

    let (reporting_proxy, reporting_handle) = create_reporting_task(reporters);

    let workers_config = WorkersConfig {
        tag,
        num_workers,
        runner_kind,
        local_results_handler: Box::new(reporting_proxy),
        worker_context: context,
        debug_native_runner: std::env::var_os("ABQ_DEBUG_NATIVE").is_some(),
        results_batch_size_hint: batch_size.get(),
    };

    tracing::debug!(
        "Workers attaching to queue negotiator {}",
        queue_negotiator.get_address()
    );

    let invoke_work = InvokeWork {
        run_id: run_id.clone(),
        batch_size_hint: batch_size,
        test_results_timeout,
    };

    let mut worker_pool = WorkersNegotiator::negotiate_and_start_pool_on_executor(
        workers_config,
        queue_negotiator,
        client_opts,
        invoke_work,
    )
    .await
    .unwrap();

    tracing::debug!("Workers attached");

    // Shut down the pool when
    //   - all its workers are done
    //   - we get a signal to shutdown
    loop {
        tokio::select! {
            () = worker_pool.wait() => {
                do_shutdown(worker_pool, reporting_handle, stdout_preferences).await;
            }
            _ = term_signals.next() => {
                do_cancellation_shutdown(worker_pool);
            }
        }
    }
}

async fn do_shutdown(
    mut worker_pool: NegotiatedWorkers,
    reporting_handle: ReportingTaskHandle,
    stdout_preferences: StdoutPreferences,
) -> ! {
    let WorkersExit {
        status,
        native_runner_info,
        manifest_generation_output,
        final_captured_outputs,
    } = worker_pool.shutdown();

    tracing::debug!("Workers shutdown");

    let finalized_reporters = reporting_handle.join().await;

    if let Some((runner, manifest_output)) = manifest_generation_output {
        print_manifest_generation_output(runner, manifest_output);
    }
    print_final_runner_outputs(final_captured_outputs);

    let completed_summary = CompletedSummary { native_runner_info };

    let (suite_result, errors) = finalized_reporters.finish(&completed_summary);

    for error in errors {
        eprintln!("{error}");
    }

    print!("\n\n");
    suite_result
        .write_short_summary_lines(&mut stdout_preferences.stdout_stream())
        .unwrap();

    // If the workers didn't fault, exit with whatever status the test suite run is at; otherwise,
    // indicate the worker fault.
    let exit_code = match status {
        WorkersExitStatus::Completed(runner_exit_code) => {
            suite_result.suggested_exit_code().max(runner_exit_code)
        }
        WorkersExitStatus::Error { errors } => {
            for error in errors {
                eprintln!("{error}");
            }
            ExitCode::ABQ_ERROR
        }
    };

    std::process::exit(exit_code.get());
}

fn do_cancellation_shutdown(worker_pool: NegotiatedWorkers) -> ! {
    let WorkersExit {
        status: _,
        native_runner_info: _,
        manifest_generation_output: _,
        final_captured_outputs,
    } = worker_pool.cancel();

    tracing::debug!("Workers cancelled");

    print_final_runner_outputs(final_captured_outputs);

    std::process::exit(ExitCode::CANCELLED.get());
}

pub(crate) fn print_manifest_generation_output(
    runner: RunnerMeta,
    manifest_output: CapturedOutput,
) {
    // NB: there is no reasonable way to surface the error at this point, since we are
    // about to shut down.
    let _opt_err = abq_reporting::output::format_manifest_generation_output(
        &mut std::io::stdout(),
        runner,
        &manifest_output,
    );
}

pub(crate) fn print_final_runner_outputs(
    final_captured_runner_outputs: Vec<(RunnerMeta, CapturedOutput)>,
) {
    for (runner, runner_out) in final_captured_runner_outputs {
        // NB: there is no reasonable way to surface the error at this point, since we are
        // about to shut down.
        let _opt_err = abq_reporting::output::format_final_runner_output(
            &mut std::io::stdout(),
            runner,
            &runner_out,
        );
    }
}
