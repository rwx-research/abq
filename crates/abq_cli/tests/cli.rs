#![cfg(test)]

use serial_test::serial;
use std::process::{ExitStatus, Output};
use std::{
    ffi::OsStr,
    io::{BufRead, BufReader},
    path::{Path, PathBuf},
    process::{Child, Command, Stdio},
};

#[cfg(feature = "test-abq-jest")]
use abq_utils::net_protocol::workers::RunId;

const WORKSPACE: &str = std::env!("ABQ_WORKSPACE_DIR");

fn var_flag_set(var: &str) -> bool {
    match std::env::var(var) {
        Ok(s) => {
            let s = s.trim();
            !s.is_empty() && s != "0" && s != "false"
        }
        Err(_) => false,
    }
}

/// Set ABQ_DEBUG_CLI_TEST=1 to write log output to files locally.
fn debug_log_for_local_run() -> bool {
    var_flag_set("ABQ_DEBUG_CLI_TESTS")
}

fn debug_log_for_ci() -> bool {
    var_flag_set("ABQ_DEBUG_CLI_TESTS_FOR_CI")
}

fn abq_binary() -> PathBuf {
    if cfg!(debug_assertions) {
        PathBuf::from(WORKSPACE).join("target/debug/abq")
    } else {
        PathBuf::from(WORKSPACE).join("target/release/abq")
    }
}

#[cfg(feature = "test-abq-jest")]
fn testdata_project(subpath: impl AsRef<Path>) -> PathBuf {
    PathBuf::from(WORKSPACE).join("testdata").join(subpath)
}

struct CmdOutput {
    stdout: String,
    stderr: String,
    exit_status: ExitStatus,
}

fn run_abq<S>(name: &str, args: impl IntoIterator<Item = S>) -> CmdOutput
where
    S: AsRef<OsStr>,
{
    run_abq_in(name, args, &std::env::current_dir().unwrap())
}

fn run_abq_in<S>(name: &str, args: impl IntoIterator<Item = S>, working_dir: &Path) -> CmdOutput
where
    S: AsRef<OsStr>,
{
    let Output {
        status,
        stdout,
        stderr,
    } = spawn_abq_in(name, args, working_dir)
        .wait_with_output()
        .expect("abq cli failed on waiting");

    CmdOutput {
        stdout: String::from_utf8(stdout).expect("abq stdout should be utf-8"),
        stderr: String::from_utf8(stderr).expect("abq stderr should be utf-8"),
        exit_status: status,
    }
}

fn spawn_abq<S>(name: &str, args: impl IntoIterator<Item = S>) -> Child
where
    S: AsRef<OsStr>,
{
    spawn_abq_in(name, args, &std::env::current_dir().unwrap())
}

fn spawn_abq_in<S>(name: &str, args: impl IntoIterator<Item = S>, working_dir: &Path) -> Child
where
    S: AsRef<OsStr>,
{
    let mut cmd = Command::new(abq_binary());

    cmd.args(args);
    cmd.current_dir(working_dir);
    cmd.stdout(Stdio::piped());
    cmd.stderr(Stdio::piped());

    if debug_log_for_local_run() {
        cmd.env("ABQ_LOGFILE", format!("{name}.debug"));
        cmd.env("ABQ_LOG", "abq=debug");
    }
    if debug_log_for_ci() {
        cmd.env("ABQ_LOGCI_WITH_PREFIX", name);
        cmd.env("ABQ_LOG", "abq=debug");
    }

    cmd.spawn().expect("abq cli failed to spawn")
}

fn find_free_port() -> u16 {
    use std::net::TcpListener;
    // TERRIBLE HACK to find a free port: bind and take that port number, assuming the OS
    // will increment later binds past the attached port, and there won't be a TOCTOU race.
    // Should be fine for testing.
    TcpListener::bind("0.0.0.0:0")
        .unwrap()
        .local_addr()
        .unwrap()
        .port()
}

fn port_active(port: u16) -> bool {
    use std::net::TcpStream;
    TcpStream::connect(("0.0.0.0", port)).is_ok()
}

const TEST_AUTH_KEY: &str = "abqs_ckoUjQN4ufq1MiUeaNllztfyqjCtuz";

/// Configuration options for how clients and servers in an abq run communicate.
/// Namely: whether they use an authentication header or not, and eventually, whether they use TLS
/// or not.
struct CSConfigOptions {
    use_auth_token: bool,
    tls: bool,
}

impl CSConfigOptions {
    fn extend_args<'a>(&self, args: &[&'a str]) -> Vec<&'a str> {
        let mut args = args.to_vec();
        if self.use_auth_token {
            args.extend(["--token", TEST_AUTH_KEY]);
        }
        if self.tls {
            args.extend(["--tls"]);
        }
        args
    }
}

/// Runs a given test with all network configuration options abq exposes.
/// That is, multiplex over whether an auth token is required.
macro_rules! test_all_network_config_options {
    (
        $(#[$cfg:meta])*
        $test_name:ident
        $run:expr
    ) => { paste::paste! {
        #[test]
        #[serial]
        $(#[$cfg])*
        #[allow(clippy::redundant_closure_call)]
        fn [<$test_name _no_auth_no_tls>]() {
            $run(&format!("{}_no_auth_no_tls", stringify!($test_name)), CSConfigOptions {
                use_auth_token: false,
                tls: false,
            })
        }

        #[test]
        #[serial]
        $(#[$cfg])*
        #[allow(clippy::redundant_closure_call)]
        fn [<$test_name _no_auth_with_tls>]() {
            $run(&format!("{}_no_auth_with_tls", stringify!($test_name)), CSConfigOptions {
                use_auth_token: false,
                tls: true,
            })
        }

        #[test]
        #[serial]
        $(#[$cfg])*
        #[allow(clippy::redundant_closure_call)]
        fn [<$test_name _with_auth_no_tls>]() {
            $run(&format!("{}_with_auth_no_tls", stringify!($test_name)), CSConfigOptions {
                use_auth_token: true,
                tls: false,
            })
        }

        #[test]
        #[serial]
        $(#[$cfg])*
        #[allow(clippy::redundant_closure_call)]
        fn [<$test_name _with_auth_with_tls>]() {
            $run(&format!("{}_with_auth_with_tls", stringify!($test_name)), CSConfigOptions {
                use_auth_token: true,
                tls: true,
            })
        }
    }};
}

test_all_network_config_options! {
    #[cfg(feature = "test-abq-jest")]
    yarn_jest_auto_workers_without_failure (|name, conf: CSConfigOptions| {
        // abq test --reporter dot (--token ...)? -- yarn jest
        let args = &["test", "--reporter", "dot"];
        let mut args = conf.extend_args(args);
        args.extend(["--", "yarn", "jest"]);
        let CmdOutput {
            stdout,
            stderr,
            exit_status,
        } = run_abq_in(
            name,
            args,
            &testdata_project("jest/npm-jest-project"),
        );

        assert!(exit_status.success());
        assert_eq!(stdout, "..");
        assert!(stderr.is_empty(), "{:?}", stderr);
    })
}

test_all_network_config_options! {
    queue_open_servers_on_specified_ports (|name, conf: CSConfigOptions| {
        let server_port = find_free_port();
        let worker_port = find_free_port();
        let negotiator_port = find_free_port();

        // abq start --bind ... --port ... --work-port ... --negotiator-port ... (--token ...)?
        let mut queue_proc = spawn_abq(name, conf.extend_args(&[
            "start",
            "--bind",
            "0.0.0.0",
            "--port",
            &server_port.to_string(),
            "--work-port",
            &worker_port.to_string(),
            "--negotiator-port",
            &negotiator_port.to_string(),
        ]));

        let queue_stdout = queue_proc.stdout.as_mut().unwrap();
        let mut queue_reader = BufReader::new(queue_stdout).lines();
        // Spin until we know the queue is UP
        loop {
            if let Some(line) = queue_reader.next() {
                let line = line.expect("line is not a string");
                if line.contains("Run the following to start workers") {
                    break;
                }
            }
        }

        assert!(port_active(server_port));
        assert!(port_active(worker_port));
        assert!(port_active(negotiator_port));

        // Must kill the queue because it sits around forever, waiting for new requests.
        queue_proc.kill().expect("queue already dead");
    })
}

test_all_network_config_options! {
    #[cfg(feature = "test-abq-jest")]
    yarn_jest_separate_queue_workers_test_without_failure (|name, conf: CSConfigOptions| {
        let server_port = find_free_port();
        let worker_port = find_free_port();
        let negotiator_port = find_free_port();

        let queue_addr = format!("0.0.0.0:{server_port}");
        let test_id = RunId::unique().to_string();

        let npm_jest_project_path = testdata_project("jest/npm-jest-project");

        // abq start --bind ... --port ... --work-port ... --negotiator-port ... (--token ...)?
        let mut queue_proc = spawn_abq(name, conf.extend_args(&[
            "start",
            "--bind",
            "0.0.0.0",
            "--port",
            &server_port.to_string(),
            "--work-port",
            &worker_port.to_string(),
            "--negotiator-port",
            &negotiator_port.to_string(),
        ]));

        let queue_stdout = queue_proc.stdout.as_mut().unwrap();
        let mut queue_reader = BufReader::new(queue_stdout).lines();
        // Spin until we know the queue is UP
        loop {
            if let Some(line) = queue_reader.next() {
                let line = line.expect("line is not a string");
                if line.contains("Run the following to start workers") {
                    break;
                }
            }
        }

        // abq work --queue-addr ... --working-dir ... (--token ...)? <test_id>
        let worker_args = &[
            "work",
            "--queue-addr",
            &queue_addr,
            "--working-dir",
            &npm_jest_project_path.display().to_string(),
        ];
        let mut worker_args = conf.extend_args(worker_args);
        worker_args.extend([test_id.as_str()]);
        let mut worker_proc = spawn_abq(&(name.to_string() + "_worker"), worker_args);

        // abq test --reporter dot --queue-addr ... --run-id ... (--token ...)? -- yarn jest
        let test_args = &[
            "test",
            "--reporter",
            "dot",
            "--queue-addr",
            &queue_addr,
            "--run-id",
            &test_id,
        ];
        let mut test_args = conf.extend_args(test_args);
        test_args.extend(["--", "yarn", "jest"]);
        let CmdOutput {
            stdout,
            stderr,
            exit_status,
        } = run_abq(&(name.to_string() + "_test"), test_args);

        assert!(exit_status.success(), "{:?}", (stdout, stderr));

        let mut lines = stdout.lines();
        assert!(lines.next().unwrap().contains("Starting test run"));
        assert_eq!(lines.next().unwrap(), "..");

        assert!(stderr.is_empty());

        let worker_exit_status = worker_proc.wait().unwrap();
        assert!(worker_exit_status.success());

        // Must kill the queue because it sits around forever, waiting for new requests.
        queue_proc.kill().expect("queue already dead");
    })
}

test_all_network_config_options! {
    #[cfg(feature = "test-abq-jest")]
    yarn_jest_auto_workers_with_failing_tests (|name, conf: CSConfigOptions| {
        // abq test --reporter dot (--token ...)? -- yarn jest
        let test_args = &[
            "test",
            "--reporter",
            "dot",
        ];
        let mut test_args = conf.extend_args(test_args);
        test_args.extend(["--", "yarn", "jest"]);
        let CmdOutput {
            stdout,
            stderr,
            exit_status,
        } = run_abq_in(
            name,
            test_args,
            &testdata_project("jest/npm-jest-project-with-failures"),
        );

        let code = exit_status.code().expect("process killed");
        assert_eq!(code, 1);

        let mut stdout_lines = stdout.lines();
        assert_eq!(stdout_lines.next().unwrap(), "F");

        assert!(stderr.is_empty(), "{:?}", stderr);
    })
}

test_all_network_config_options! {
    #[cfg(feature = "test-abq-jest")]
    yarn_jest_separate_queue_workers_with_failing_tests (|name, conf: CSConfigOptions| {
        let server_port = find_free_port();
        let worker_port = find_free_port();
        let negotiator_port = find_free_port();

        let queue_addr = format!("0.0.0.0:{server_port}");
        let test_id = RunId::unique().to_string();

        let npm_jest_project_path = testdata_project("jest/npm-jest-project-with-failures");

        // abq start --bind ... --port ... --work-port ... --negotiator-port ... (--token ...)?
        let mut queue_proc = spawn_abq(name, conf.extend_args(&[
                "start",
                "--bind",
                "0.0.0.0",
                "--port",
                &server_port.to_string(),
                "--work-port",
                &worker_port.to_string(),
                "--negotiator-port",
                &negotiator_port.to_string(),
        ]));

        let queue_stdout = queue_proc.stdout.as_mut().unwrap();
        let mut queue_reader = BufReader::new(queue_stdout).lines();
        // Spin until we know the queue is UP
        loop {
            if let Some(line) = queue_reader.next() {
                let line = line.expect("line is not a string");
                if line.contains("Run the following to start workers") {
                    break;
                }
            }
        }

        // abq work --queue-addr ... --working-dir ... (--token ...)? <test_id>
        let worker_args = &[
            "work",
            "--queue-addr",
            &queue_addr,
            "--working-dir",
            &npm_jest_project_path.display().to_string(),
        ];
        let mut worker_args = conf.extend_args(worker_args);
        worker_args.extend([test_id.as_str()]);
        let mut worker_proc = spawn_abq(&(name.to_string() + "_workers"), worker_args);

        // abq test --reporter dot --queue-addr ... --run-id ... (--token ...)? -- yarn jest
        let test_args = &[
            "test",
            "--reporter",
            "dot",
            "--queue-addr",
            &queue_addr,
            "--run-id",
            &test_id,
        ];
        let mut test_args = conf.extend_args(test_args);
        test_args.extend(["--", "yarn", "jest"]);
        let CmdOutput {
            stdout,
            stderr,
            exit_status,
        } = run_abq(&(name.to_string() + "_test"), test_args);

        // The `abq test` process should exit with a failure
        assert!(!exit_status.success(), "{:?}", (stdout, stderr));

        let mut lines = stdout.lines();
        assert!(lines.next().unwrap().contains("Starting test run"));
        assert_eq!(lines.next().unwrap(), "F");

        assert!(stderr.is_empty());

        // The `abq work` process should also exit with a failure, corresponding to having
        // witnessed a test failure in the run.
        let worker_exit_status = worker_proc.wait().unwrap();
        assert!(!worker_exit_status.success());

        // Must kill the queue because it sits around forever, waiting for new requests.
        queue_proc.kill().expect("queue already dead");
    })
}

test_all_network_config_options! {
    healthcheck_queue_success (|name, conf: CSConfigOptions| {
        let server_port = find_free_port();
        let work_port = find_free_port();
        let negotiator_port = find_free_port();

        let queue_addr = format!("0.0.0.0:{server_port}");
        let work_scheduler_addr = format!("0.0.0.0:{work_port}");
        let negotiator_addr = format!("0.0.0.0:{negotiator_port}");

        // abq start --bind ... --port ... --work-port ... --negotiator-port ... (--token ...)?
        let mut queue_proc = spawn_abq(name, conf.extend_args(&[
            "start",
            "--bind",
            "0.0.0.0",
            "--port",
            &server_port.to_string(),
            "--work-port",
            &work_port.to_string(),
            "--negotiator-port",
            &negotiator_port.to_string(),
        ]));

        let queue_stdout = queue_proc.stdout.as_mut().unwrap();
        let mut queue_reader = BufReader::new(queue_stdout).lines();
        // Spin until we know the queue is UP
        loop {
            if let Some(line) = queue_reader.next() {
                let line = line.expect("line is not a string");
                if line.contains("Run the following to start workers") {
                    break;
                }
            }
        }

        // Check health
        // abq health --queue ... --work-scheduler ... --negotiator ... (--token ...)?
        let CmdOutput {
            stdout,
            stderr,
            exit_status,
        } = run_abq(&(name.to_string() + "_health"), conf.extend_args(&[
            "health",
            "--queue",
            &queue_addr,
            "--work-scheduler",
            &work_scheduler_addr,
            "--negotiator",
            &negotiator_addr,
        ]));

        assert!(exit_status.success(), "{}\n{}", stdout, stderr);
        assert!(stdout.is_empty());
        assert!(stderr.is_empty());

        queue_proc.kill().expect("queue already dead");
    })
}

test_all_network_config_options! {
    healthcheck_queue_failure (|name, conf: CSConfigOptions| {
        let server_port = find_free_port();
        let work_port = find_free_port();
        let negotiator_port = find_free_port();

        let queue_addr = format!("0.0.0.0:{server_port}");
        let work_scheduler_addr = format!("0.0.0.0:{work_port}");
        let negotiator_addr = format!("0.0.0.0:{negotiator_port}");

        // abq start --bind ... --port ... --work-port ... --negotiator-port ... (--token ...)?
        let mut queue_proc = spawn_abq(name, conf.extend_args(&[
            "start",
            "--bind",
            "0.0.0.0",
            "--port",
            &server_port.to_string(),
            "--work-port",
            &work_port.to_string(),
            "--negotiator-port",
            &negotiator_port.to_string(),
        ]));

        queue_proc.kill().expect("queue already dead");

        // Check health
        // abq health --queue ... --work-scheduler ... --negotiator ... (--token ...)?
        let CmdOutput {
            stdout,
            stderr,
            exit_status,
        } = run_abq(&(name.to_string() + "_health"), conf.extend_args(&[
            "health",
            "--queue",
            &queue_addr,
            "--work-scheduler",
            &work_scheduler_addr,
            "--negotiator",
            &negotiator_addr,
        ]));

        assert_eq!(exit_status.code().unwrap(), 1);
        assert_eq!(
            stdout,
            format!(
                r#"Queue at {queue_addr}: unhealthy
Work scheduler at {work_scheduler_addr}: unhealthy
Negotiator at {negotiator_addr}: unhealthy
"#
            )
        );
        assert!(stderr.is_empty());
    })
}

#[test]
#[serial]
fn work_no_queue_addr_or_api_key() {
    // Spawn worker without a queue addr or api key
    // abq work --working-dir . run-id
    let CmdOutput {
        stdout,
        stderr,
        exit_status,
    } = run_abq(
        "work_no_queue_addr_or_api_key",
        ["work", "--working-dir", ".", "run-id"],
    );

    assert_eq!(exit_status.code().unwrap(), 2);
    assert!(stdout.is_empty());
    insta::assert_snapshot!(stderr);
}
