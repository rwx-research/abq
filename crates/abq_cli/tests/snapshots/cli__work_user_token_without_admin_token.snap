---
source: crates/abq_cli/tests/cli.rs
expression: stderr
---
error: the following required arguments were not provided:
  --admin-token <ADMIN_TOKEN>

Usage: abq start --user-token <USER_TOKEN> --admin-token <ADMIN_TOKEN>

For more information, try '--help'.

